sycl/test-e2e/VirtualMem/vector_with_virtual_mem.cpp

```diff
@@ -0,0 +1,195 @@
+
+  ~VirtualVector() {
+    // Free all mapped ranges.
+    for (const VirtualAddressRange &VARange : MVARanges) {
+      syclext::unmap(VARange.Ptr, VARange.Size, MContext);
+      syclext::free_virtual_mem(VARange.Ptr, VARange.Size, MContext);
+    }
+    // Physical memory allocations will be freed when the physical_mem objects
+    // die with MPhysicalMems.
+  }
+
+  void reserve(size_t NewSize) {
+    // If we already have more memory than required, we can return.
+    size_t NewByteSize = sizeof(T) * NewSize;
+    if (NewByteSize <= MByteSize) {
+      MSize = NewSize;
+      return;
+    }
+
+    // Align the size by the minimum granularity.
+    size_t AlignedNewByteSize = AlignByteSize(NewByteSize);
+    size_t AlignedNewVARangeSize = AlignedNewByteSize - MByteSize;
+
+    // Try to reserve virtual memory at the end of the existing one.
+    void *CurrentEnd = reinterpret_cast<char *>(MBasePtr) + MByteSize;
+    void *NewVAPtr = syclext::reserve_virtual_mem(
+        CurrentEnd, AlignedNewVARangeSize, MContext);
+
+    // If we failed to get a ptr to the end of the current range, we need to
+    // recreate the whole range.
+    if (CurrentEnd && NewVAPtr != CurrentEnd) {
+      // First we need to free the virtual address range we just reserved.
+      syclext::free_virtual_mem(NewVAPtr, AlignedNewVARangeSize, MContext);
+
+      // Recreate the full range and update the new VA ptr. CurrentEnd is no
+      // longer valid after this call.
+      NewVAPtr = RecreateAddressRange(AlignedNewByteSize);
+    } else {
+      // Otherwise we need to register the new range.
+      MVARanges.emplace_back(NewVAPtr, AlignedNewVARangeSize);
+
+      // If there was no base pointer previously, this is now the new base.
+      if (!MBasePtr)
+        MBasePtr = reinterpret_cast<T *>(NewVAPtr);
+    }
+
+    // Create new physical memory allocation and map the new range to it.
+    syclext::physical_mem NewPhysicalMem{MDevice, MContext,
+                                         AlignedNewVARangeSize};
+    NewPhysicalMem.map(NewVAPtr, AlignedNewVARangeSize, 0,
+                       sycl::access_mode::read_write);
+    MPhysicalMems.push_back(std::move(NewPhysicalMem));
+
+    // Update the byte size of the vector.
+    MSize = NewSize;
+    MByteSize = AlignedNewByteSize;
+  }
+
+  size_t size() const noexcept { return MSize; }
+  T *data() const noexcept { return MBasePtr; }
+
+private:
+  size_t AlignByteSize(size_t UnalignedByteSize) const {
+    size_t Granularity = syclext::get_minimum_mem_granularity(
+        UnalignedByteSize, MDevice, MContext);
+    return ((UnalignedByteSize + Granularity - 1) / Granularity) * Granularity;
+  }
+
+  void *RecreateAddressRange(size_t AlignedNewByteSize) {
+    // Reserve the full range.
+    void *NewFullVAPtr =
+        syclext::reserve_virtual_mem(AlignedNewByteSize, MContext);
+
+    // Unmap the old virtual address in its entirety.
+    syclext::unmap(MBasePtr, MByteSize, MContext);
```

Comment:
Testing my understanding of this code ...

I think the address range `[MBasePtr, MBasePtr + MBytesSize]` could represent several different mapped `physical_mem` objects because the code in `reserve` attempts to incrementally map new physical memory at the end of the previous allocation.  If that is successful, then this range could represent several different mapped physical memory regions.

Does that mean:

* The call here to `unmap` is unmapping all physical memory regions in that address range?  This was not my impression of the semantics, but maybe I'm not reading the spec correctly, or
* The code has a bug here, and it should have a loop with separate calls to `unmap`

?

Reply:
This could potentially unmap multiple virtual address ranges from multiple `physical_mem`, as long as the virtual address ranges are contiguous. The CUDA API ([cuMemUnmap](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__VA.html#group__CUDA__VA_1gfb50aac00c848fd7087e858f59bf7e2a)) allows this, but reading the L0 specification ([zeVirtualMemUnmap](https://spec.oneapi.io/level-zero/latest/core/api.html#zevirtualmemunmap)) again, it seems to specify that
> Unmaps pages in virtual address space from pages from a physical memory object.
which suggests it may not be allowed to unmap a full range mapped over multiple physical memory objects. @jandres742 - Do you know if this is intentional?

Reply:
@steffenlarsen : that seems to be more of a miss. There's no reason why from the point of view of the spec that zeVirtualMemUnmap couldn't unmap from different physical locations. So behavior of L0 and CUDA here aligns.